pkg/logql/log/distinct.go

```diff
@@ -0,0 +1,37 @@
+package log
+
+type distinctFilter struct {
+	datas  map[string]map[string]int
+	labels []string
+}
+
+func NewDistinctFilter(labels []string) (Stage, error) {
+	datas := make(map[string]map[string]int, 0)
+	for _, label := range labels {
+		datas[label] = make(map[string]int, 0)
+	}
+	return &distinctFilter{
+		labels: labels,
+		datas:  datas,
+	}, nil
+}
+func (r *distinctFilter) Process(_ int64, line []byte, lbs *LabelsBuilder) ([]byte, bool) {
+	isPass := false
+	for _, label := range r.labels {
+		val, ok := lbs.Get(label)
+		if !ok {
+			return line, false
+		}
+		_, ok = r.datas[label][val]
```

Comment:
If the label does not exist, it will be discarded, and the new label is based on the order of the pipeline. Such as the following example

log
`{"a":"1", "b":2}`

logql 1 :
`{app="buy2"} | json | distinct a
`
logql 2 :
`{app="buy2"} | distinct a | json`

logql 1 can get log. logql 2 will not get log.

Reply:
Hm, but you'd still have a nil pointer right?

Reply:
good point.This code does look dangerous.
No, each label has been initialized in the `NewDistinctFilter` () function.
```go
func NewDistinctFilter(labels []string) (Stage, error) {
	datas := make(map[string]map[string]int, 0)
	for _, label := range labels {
		datas[label] = make(map[string]int, 0)
	}
	...
}
```
If `distinctFilter` struct initialized through `NewDistinctFilter`(), the `distinctFilter` struct will never get nil point. But if we directly build a `distinctFilter` struct with &`distinctFilter`{}, it is possible that the nil point you mentioned will appear.

another way. 
![image](https://user-images.githubusercontent.com/9583245/233519713-dfbb2db9-e6cb-4b57-a2eb-c198d3e7fd8e.png)
this for performance, it will not need to make a judgment like this for each log.

```go
 _, ok = r.datas[label]
if !ok {
r.datas[label] = make(map[string]int, 0)
}
```

Reply:
Is `NewDestinctFilter` called for each line? How is preserving the state?

Reply:
`NewDestinctFilter` is generated for each `logql`, not for each line. So the `state` in `NewDestinctFilter` can be maintained continuously in one logql. The screenshot in the PR can also show that this `state` is valid

![image](https://user-images.githubusercontent.com/9583245/235030902-9b3b0829-1208-462e-b05b-bb11fed84031.png)

Reply:
Thanks for your patience. I still don't understand how `data[label]` can be not nil if the label not known when `NewDestinctFilter` is called. I'll try out.

Reply:
Ah, no I get it. The labels are not the labels seen in the line but the ones passed to `distinct` :see_no_evil: